Sources/System/FilePath/FilePathSyntax.swift

```diff
@@ -0,0 +1,353 @@
+// Query API
+// @available(...)
+extension FilePath {
+  /// Returns true if this path uniquely identifies the location of
+  /// a file without reference to an additional starting location.
+  ///
+  /// On Unix platforms, absolute paths begin with a `/`. `isAbsolute` is equivalent
+  /// to `root != nil`.
+  ///
+  /// On Windows, absolute paths are fully qualified paths. UNC paths and device paths
+  /// are always absolute. Traditional DOS paths are absolute if they begin with a volume or drive
+  /// followed by a `:` and a separator.
+  ///
+  /// NOTE: This does not perform shell expansion or substitute
+  /// environment variables; paths beginning with `~` are considered relative.
+  ///
+  /// Examples:
+  /// * Unix:
+  ///   * `/usr/local/bin`
+  ///   * `/tmp/foo.txt`
+  ///   * `/`
+  /// * Windows:
+  ///   * `C:\Users\`
+  ///   * `\\?\UNC\server\share\bar.exe`
+  ///   * `\\server\share\bar.exe`
+  ///
+  /// TODO: We should add API to system to access environment variables
+  /// and perform expansion/substituion, but that's deferred for now. Should
+  /// we mention that?
+  public var isAbsolute: Bool {
+    self.root?._rootIsAbsolute ?? false
+  }
+
+  /// Returns true if this path is not absolute (see absolute)
+  ///
+  /// Examples:
+  /// * Unix:
+  ///   * `~/bar`
+  ///   * `tmp/foo.txt`
+  /// * Windows:
+  ///   * `bar\baz`
+  ///   * `C:Users\`
+  ///   * `\Users`
+  public var isRelative: Bool { !isAbsolute }
+
+  /// A textual representation of the file path using the generic separator (`/`).
+  ///
+  /// On Unix, this is equivalent to `description`
+  ///
+  /// TODO: Should `dump()` use this? Also, should dump escape invalid Unicode?
+  public var genericDescription: String {
+    fatalError()
+  }
+
+  /// Returns whether `self` is a prefix of `other`
+  public func starts(with other: FilePath) -> Bool {
+    fatalError()
+  }
+
+  /// Returns whether `self` is a suffix of `other`
+  public func ends(with other: FilePath) -> Bool {
+    fatalError()
+  }
+
+  /// Whether this path is empty
+  public var isEmpty: Bool { self.storage == [0] }
+}
+
+
+// Decompose a path into its parts
+//
+//   path = root? relative-path?
+//   root = root-name? root-directory?
+//   root-directory = separator
+//   root-name (Windows only) = disk | UNC | device
+extension FilePath {
+  /// Returns the root directory of a path if there is one, otherwise `nil`
+  ///
+  /// On Unix, this will return the leading `/` if the path is absolute
```

Comment:
I think this makes a lot of sense.  For TSCs/SwiftPM's `AbsolutePath` we debated whether the parent directory of the root path should be nil, meaning that every directory might potentially not have a parent, but that gets really unergonomic at the call sites.  Framing this in terms of what the root is, and having the root have itself as the root, makes a ton of sense I think.